### PR TITLE
Feature/get missing alias json exception 301

### PR DESCRIPTION
**Pull request by szczepanczykd**
### Description
Handling String error response from /_alias/not_existing_alias endpoint.

Add errorDeserializer to SimpleEndpoint constructor -> and use ErrorResponse._DESERIALIZER in every instance (except the GetAliasRequest._ENDPOINT) -> maybe two constructors would be better (one with default errorDeserializer)?

Created OpenSearchExceptionFactory -> thrown OpenSearchException is untouched.

API error response for GET https://localhost:9200/_alias/not_existing_alias (ErrorStringResponse format)
image

API error response for GET https://localhost:9200/not_existing_index/_search (ErrorResponse format)
image

### Issues Resolved
https://github.com/opensearch-project/opensearch-java/issues/301

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:9]: New code or copied from ES?

If new, remove this part. 
If from ES, confirm that it's APLv2 please.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+//----------------------------------------------------
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:33]: Was it actually generated?

File: java-client/src/main/java/org/opensearch/client/transport/endpoints/SimpleEndpoint.java [hunk 3]
```diff
@@ -60,21 +60,41 @@ public static <T> Function<T, Map<String, String>> emptyMap() {
     private final Function<RequestT, Map<String, String>> headers;
     private final boolean hasRequestBody;
     private final JsonpDeserializer<ResponseT> responseParser;
+    private final JsonpDeserializer<ErrorResponseT> errorDeserializer;
+
+    public SimpleEndpoint(
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/endpoints/SimpleEndpoint.java:65]: Let's write one constructor with fewer parameters to reuse the other with more parameters?

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+package org.opensearch.client.opensearch._types;
+
+public class OpenSearchExceptionFactory {
+
+    private OpenSearchExceptionFactory() {
+        //should be empty
+    }
+
+    public static <ErrorT> OpenSearchException createException(ErrorT error) {
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java:41]: Add tests for the factory?

**Review by dblock — CHANGES_REQUESTED** [review]
Looks OK to me. Made some smallish suggestions.

Should we open an issue in OpenSearch to not return a different response format in 3.0? This seems like an odd choice server-side. Are there other places where the server does this?

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java (hunk 2 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:33]: No, it's not.
Removed.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java (hunk 1 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:9]: Removed.

File: java-client/src/main/java/org/opensearch/client/transport/endpoints/SimpleEndpoint.java (hunk 3 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/transport/endpoints/SimpleEndpoint.java:65]: Done.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java (hunk 4 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java:22]: Done

**Comment by szczepanczykd** [comment]
> Looks OK to me. Made some smallish suggestions.
> 
> Should we open an issue in OpenSearch to not return a different response format in 3.0? This seems like an odd choice server-side. Are there other places where the server does this?

I think yes, it seems it is only one case:
https://github.com/opensearch-project/OpenSearch/blob/8e0edee3d403e5f2339d176b800e9e5df1372692/server/src/main/java/org/opensearch/rest/action/admin/indices/RestGetAliasesAction.java

Lines: [160, 162]

Thanks for the review :)

**Comment by dblock** [comment]
> > Looks OK to me. Made some smallish suggestions.
> > Should we open an issue in OpenSearch to not return a different response format in 3.0? This seems like an odd choice server-side. Are there other places where the server does this?
> 
> I think yes, it seems it is only one case: https://github.com/opensearch-project/OpenSearch/blob/8e0edee3d403e5f2339d176b800e9e5df1372692/server/src/main/java/org/opensearch/rest/action/admin/indices/RestGetAliasesAction.java
> 
> Lines: [160, 162]
> 
> Thanks for the review :)

Will you please do it?

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java (hunk 1 above)
Comment (dblock) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:9]: Remove the `Modifications Copyright OpenSearch...` part too. We only need the SPDX header.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+import org.opensearch.client.json.JsonpSerializable;
+import org.opensearch.client.json.ObjectBuilderDeserializer;
+import org.opensearch.client.json.ObjectDeserializer;
+import org.opensearch.client.util.ApiTypeHelper;
+import org.opensearch.client.util.ObjectBuilder;
+import org.opensearch.client.util.ObjectBuilderBase;
+
+import java.util.function.Function;
+
+// typedef: _types.ErrorSimpleResponseBase
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:29]: Remove?

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    // ---------------------------------------------------------------------------------------------
+
+    /**
+     * Json deserializer for {@link ErrorStringResponse}
+     */
+    public static final JsonpDeserializer<ErrorStringResponse> _DESERIALIZER = ObjectBuilderDeserializer.lazy(Builder::new,
+            ErrorStringResponse::setupErrorResponseDeserializer);
+
+    protected static void setupErrorResponseDeserializer(ObjectDeserializer<ErrorStringResponse.Builder> op) {
+
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:142]: Nit: Not sure what the convention in this project is, let's remove the extra empty lines in the beginning/end of functions.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.client.opensearch._types;
+
+public class OpenSearchExceptionFactory {
+
+    private OpenSearchExceptionFactory() {
+        //should be empty
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java:19]: Remove

**Review by dblock — CHANGES_REQUESTED** [review]
I have some nitpicky asks, the only must have is the license header update, we don't need the "modifications ..." part.

I think the rest of the code uses two spaces instead of 4/tabs? Mind checking and matching that. We should have some style checker/autoformatter do this and fail CI otherwise, too.

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java (hunk 7 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/OpenSearchExceptionFactory.java:19]: Done

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java (hunk 5 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:29]: Done

File: java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java (hunk 1 above)
Comment (szczepanczykd) [reply java-client/src/main/java/org/opensearch/client/opensearch/_types/ErrorStringResponse.java:9]: Done

**Comment by szczepanczykd** [comment]
> I have some nitpicky asks, the only must have is the license header update, we don't need the "modifications ..." part.
> 
> I think the rest of the code uses two spaces instead of 4/tabs? Mind checking and matching that. We should have some style checker/autoformatter do this and fail CI otherwise, too.

The tests are using 4 spaces.
Rest of the code uses tabs.

**Comment by szczepanczykd** [comment]
> > > Looks OK to me. Made some smallish suggestions.
> > > Should we open an issue in OpenSearch to not return a different response format in 3.0? This seems like an odd choice server-side. Are there other places where the server does this?
> > 
> > 
> > I think yes, it seems it is only one case: https://github.com/opensearch-project/OpenSearch/blob/8e0edee3d403e5f2339d176b800e9e5df1372692/server/src/main/java/org/opensearch/rest/action/admin/indices/RestGetAliasesAction.java
> > Lines: [160, 162]
> > Thanks for the review :)
> 
> Will you please do it?

Done: https://github.com/opensearch-project/OpenSearch/issues/6904

**Comment by dblock** [comment]
Re-running flaky tests. Someone should look at those, wink wink :)

**Review by dblock — APPROVED** [review]

**Comment by reta** [comment]
> Re-running flaky tests. Someone should look at those, wink wink :)

There is a fix suggested already https://github.com/opensearch-project/opensearch-java/pull/421

**Comment by dblock** [comment]
@reta I just merged that. Shall we just merge this one?

**Comment by reta** [comment]
> @reta I just merged that. Shall we just merge this one?

@dblock the main is broken now (after https://github.com/opensearch-project/OpenSearch/pull/6875), fixing it now

**Comment by VachaShah** [comment]
Hi @szczepanczykd, apologies for the delay! Can you merge the latest changes from main and fix the conflicts? We can get this merged once the CI is green.

**Comment by szczepanczykd** [comment]
Hi :)

Sure, I will do it today, around 18 UTC.

**Comment by szczepanczykd** [comment]
> Hi @szczepanczykd, apologies for the delay! Can you merge the latest changes from main and fix the conflicts? We can get this merged once the CI is green.

@VachaShah done :)

**Comment by reta** [comment]
@dblock @szczepanczykd @VachaShah I think this change is good but it does not follow the way we manage such API elements (when multiple forms are possible).  Generally, it is done as a union (see for example https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/opensearch/_types/Script.java#L164 but there are much more).

The way I would see it working is:
 - in both cases, the error is the same class ErrorResponse
 - but depending on the type (Object or String - solved by `UnionDeserializer`), it is either a full one or just `ErrorCause` with message populated 

WDYT?

**Comment by szczepanczykd** [comment]
Sounds good.
Let's try it.

**Comment by szczepanczykd** [comment]
@VachaShah @reta 
created a new PR with a simpler approach to this issue:
https://github.com/opensearch-project/opensearch-java/pull/476

**Pull request closed without merging**